dspace-swordv2/src/main/java/org/dspace/sword2/VersionManager.java

```diff
@@ -10,75 +10,90 @@
 			throws SQLException, AuthorizeException, IOException
 	{
 		boolean keep = ConfigurationManager.getBooleanProperty("swordv2-server", "versions.keep");
-		Bundle[] bundles = item.getBundles(name);
+		List<Bundle> bundles = item.getBundles();
 		for (Bundle b : bundles)
 		{
-			this.removeBundle(item, b, keep);
+			if (name.equals(b.getName()))
+			{
+				this.removeBundle(context, item, b, keep);
+			}
 		}
 	}
 
-	public void removeBundle(Item item, Bundle source)
+	public void removeBundle(Context context, Item item, Bundle source)
 			throws SQLException, AuthorizeException, IOException
 	{
 		boolean keep = ConfigurationManager.getBooleanProperty("swordv2-server", "versions.keep");
-		this.removeBundle(item, source, keep);
+		this.removeBundle(context, item, source, keep);
 	}
 
-	public void removeBundle(Item item, Bundle source, boolean archive)
+	public void removeBundle(Context context, Item item, Bundle source, boolean archive)
 			throws SQLException, AuthorizeException, IOException
 	{
         // archive the bundle contents if desired
 		if (archive)
 		{
-			this.archiveBundle(item, source);
+			this.archiveBundle(context, item, source);
 		}
 
         // remove all the bitstreams from the bundle
-		for (Bitstream bitstream : source.getBitstreams())
+		List<BundleBitstream> bundleBitstreams = source.getBitstreams();
+		for (BundleBitstream bundleBitstream : bundleBitstreams)
 		{
-			source.removeBitstream(bitstream);
+			bundleService.removeBitstream(context, source, bundleBitstream.getBitstream());
 		}
 
         // delete the bundle itself
-        item.removeBundle(source);
+		itemService.removeBundle(context, item, source);
 	}
 
-	public void removeBitstream(Item item, Bitstream bitstream)
+	public void removeBitstream(Context context, Item item, Bitstream bitstream)
 			throws SQLException, AuthorizeException, IOException
 	{
 		boolean keep = ConfigurationManager.getBooleanProperty("swordv2-server", "versions.keep");
-		this.removeBitstream(item, bitstream, keep);
+		this.removeBitstream(context, item, bitstream, keep);
 	}
 
-	public void removeBitstream(Item item, Bitstream bitstream, boolean keep)
+	public void removeBitstream(Context context, Item item, Bitstream bitstream, boolean keep)
 			throws SQLException, AuthorizeException, IOException
 	{
 		Bundle exempt = null;
 		if (keep)
 		{
-			exempt = this.archiveBitstream(item, bitstream);
+			exempt = this.archiveBitstream(context, item, bitstream);
 		}
 
-		Bundle[] bundles = bitstream.getBundles();
-		for (Bundle bundle : bundles)
+		List<BundleBitstream> bundleBitstreams = bitstream.getBundles();
+		for (BundleBitstream bundleBitstream : bundleBitstreams)
 		{
-			if (exempt != null && bundle.getID() != exempt.getID())
+			if (exempt != null && bundleBitstream.getBundle().getID() != exempt.getID())
 			{
-				bundle.removeBitstream(bitstream);
+				bundleService.removeBitstream(context, bundleBitstream.getBundle(), bitstream);
```

Comment:
I think that we will get an "ConcurrentModificationException" here. Because the "BundleBitstream" list is persisted, so when we delete an item from a list we are iterating it will alter the list.
The best way to fix this is by using an iterator (see comment above for example)